ntoskrnl/fsrtl/fastio.c

```diff
@@ -1661,10 +1661,54 @@ FsRtlReleaseFile(IN PFILE_OBJECT FileObject)
     PFSRTL_COMMON_FCB_HEADER FcbHeader;
     PDEVICE_OBJECT DeviceObject;
     PFAST_IO_DISPATCH FastDispatch;
+    PEXTENDED_DRIVER_EXTENSION DriverExtension;
+    PFS_FILTER_CALLBACKS FilterCallbacks;
 
     /* Get Device Object and Fast Calls */
     FcbHeader = (PFSRTL_COMMON_FCB_HEADER)FileObject->FsContext;
     DeviceObject = IoGetRelatedDeviceObject(FileObject);
+
+    DriverExtension = (PEXTENDED_DRIVER_EXTENSION)DeviceObject->DriverObject->DriverExtension;
+    FilterCallbacks = DriverExtension->FsFilterCallbacks;
+
+    /* Check if Filter Callbacks are supported */
+    if (FilterCallbacks && FilterCallbacks->PreReleaseForSectionSynchronization)
+    {
+        NTSTATUS Status;
+        PVOID CompletionContext;
+
+        FS_FILTER_CALLBACK_DATA CbData;
+
+        RtlZeroMemory(&CbData, sizeof(CbData));
+
+        CbData.SizeOfFsFilterCallbackData = sizeof(CbData);
+        CbData.Operation = FS_FILTER_RELEASE_FOR_SECTION_SYNCHRONIZATION;
+        CbData.DeviceObject = DeviceObject;
+        CbData.FileObject = FileObject;
+
+        Status = FilterCallbacks->PreReleaseForSectionSynchronization(&CbData, &CompletionContext);
+        if (!NT_SUCCESS(Status))
+        {
+            FsRtlExitFileSystem();
+            return;
+        }
+
+        /* Should we do something in-between ? */
+
+        if (FilterCallbacks->PostReleaseForSectionSynchronization)
+        {
+            FilterCallbacks->PostReleaseForSectionSynchronization(&CbData, Status, CompletionContext);
+        }
+
+        /* Return here when the status is based on the synchonization type and write access to the file */
```

Comment:
Now, with this PR, I wonder whether you really need these returns when Status takes these specific values.
Could you retest the AVs you were testing without these? (you would need to disable these returns in all your callback handlings.)

Reply:
Nope, does not works for me. :slightly_frowning_face: Can't test my AVs, because it does not even boot: hangs on e1000 driver loading:
![VirtualBox_ReactOS_30_08_2024_21_53_54](https://github.com/user-attachments/assets/59680833-f7fd-4b3c-b169-806124bc2b23)
Nothing suspicious in the log: no any assert/exception/bugcheck hit, just stucks at loading.
Seems I need to return something there still, but not this probably...
With AMD PCNET III instead of E1000 enabled, it hangs on loading VBoxVideo/AC97 instead:
![VirtualBox_ReactOS_30_08_2024_21_58_05](https://github.com/user-attachments/assets/da3791cc-0ae8-481f-aa93-f93624dd63db)

Reply:
have you checked the debug logs?

Reply:
Yeah, nothing suspicious there:
```
BOOT DRIVERS LOADED
(ntoskrnl/io/pnpmgr/pnpres.c:789) IopInitiatePnpIrp(IRP_MN_FILTER_RESOURCE_REQUIREMENTS) failed
(ntoskrnl/io/pnpmgr/pnpres.c:789) IopInitiatePnpIrp(IRP_MN_FILTER_RESOURCE_REQUIREMENTS) failed
(drivers/storage/port/scsiport/ioctl.c:542) unknown ioctl code: 0x66001B
(ntoskrnl/mm/ARM3/sysldr.c:170) Loading: \SystemRoot\System32\drivers\sysaudio.sys at F6FA7000 with f pages
(ntoskrnl/mm/ARM3/sysldr.c:170) Loading: \SystemRoot\system32\drivers\kmixer.sys at F6E2A000 with 17d pages
(drivers/wdm/audio/filters/kmixer/kmixer.c:106) KMixer.sys loaded
(drivers/ksfilter/swenum/swenum.c:107) SwDispatchPnp ChildDevice 1 Request d
(drivers/ksfilter/swenum/swenum.c:107) SwDispatchPnp ChildDevice 1 Request 0
(drivers/ksfilter/ks/swenum.c:812) KspStartBusDevice Name \Device\KSENUM00000001 DeviceName {a7c7a5b0-5af3-11d1-9ced-00a024bf0407}&{9b365890-165f-11d0-a195-0020afd156e4} Instance {9b365890-165f-11d0-a195-0020afd156e4} Started
(drivers/ksfilter/swenum/swenum.c:107) SwDispatchPnp ChildDevice 1 Request 9
(drivers/ksfilter/swenum/swenum.c:107) SwDispatchPnp ChildDevice 1 Request 14
(drivers/ksfilter/swenum/swenum.c:107) SwDispatchPnp ChildDevice 1 Request 7
(ntoskrnl/mm/ARM3/sysldr.c:170) Loading: \SystemRoot\System32\drivers\processr.sys at F6E21000 with 9 pages
(ntoskrnl/mm/ARM3/sysldr.c:170) Loading: \SystemRoot\System32\drivers\VBoxVideo.sys at F6DE4000 with 3d pages
(ntoskrnl/mm/ARM3/sysldr.c:170) Loading: \SystemRoot\System32\drivers\VIDEOPRT.SYS at F6DB6000 with 2e pages
IPRT: RTMpPoke => rtMpPokeCpuUsingDpc
(ntoskrnl/io/pnpmgr/pnpres.c:648) Resource conflict: IRQ (0xa 0xa vs. 0xa 0xa)
(ntoskrnl/io/pnpmgr/pnpres.c:112) Satisfying memory requirement with 0xa0000 (length: 0x20000)
(ntoskrnl/io/pnpmgr/pnpres.c:648) Resource conflict: IRQ (0xa 0xa vs. 0xa 0xa)
(ntoskrnl/kdbg/kdb_symbols.c:231) Trying \SystemRoot\System32\drivers\VBoxVideo.sys
(ntoskrnl/mm/ARM3/sysldr.c:170) Loading: \SystemRoot\system32\drivers\ac97.sys at F6D5D000 with 45 pages
```
[debug.log](https://github.com/user-attachments/files/16821663/debug.log)

Reply:
My investigations proved that the current code is correct instead, so keeping it as-is. :slightly_smiling_face: